clang/lib/Analysis/FlowSensitive/Models/ReverseNullModel.cpp

```diff
@@ -0,0 +1,723 @@
+  }
+
+  if (Env.flowConditionImplies(Val->formula()))
+    return SR::True;
+  
+  if (Env.flowConditionImplies(Env.arena().makeNot(Val->formula())))
+    return SR::False;
+
+  return SR::Unknown;
+}
+
+inline BoolValue &getVal(llvm::StringRef Name, Value &RootValue) {
+    auto *Value = cast_or_null<BoolValue>(RootValue.getProperty(Name));
+    assert(Value && "boolean value was not created");
+    return *Value;
+}
+
+void initializeRootValue(Value &RootValue, Environment &Env) {
+
+    Arena &A = Env.arena();
+
+    BoolValue &IsNull = A.makeAtomValue();
+    BoolValue &IsNonnull = A.makeAtomValue();
+    RootValue.setProperty(kIsNull, IsNull);
+    RootValue.setProperty(kIsNonnull, IsNonnull);
+
+    Env.addToFlowCondition(A.makeOr(IsNull.formula(), IsNonnull.formula()));
+}
+
+void setLValue(const Expr &E, Value &Val, Environment &Env) {
+  StorageLocation *SL = Env.getStorageLocation(E);
+  if (!SL) {
+    SL = &Env.createStorageLocation(E);
+    Env.setStorageLocation(E, *SL);
+  }
+
+  Env.setValue(*SL, Val);
+}
+
+void setUnknownValue(const Expr &E, Value &Val, Environment &Env) {
+  if (E.isLValue()) {
+    setLValue(E, Val, Env);
+  } else if (E.isPRValue()) {
+    Env.setValue(E, Val);
+  } else {
+    llvm_unreachable("all value cases covered");
+  }
+}
+
+Value *getValue(const Expr &Var, Environment &Env) {
+    if (auto *EnvVal = Env.getValue(Var)) {
+        // FIXME: Hack
+        if (!EnvVal->getProperty(kIsNull) && !EnvVal->getProperty(kIsNonnull)) {
+            initializeRootValue(*EnvVal, Env);
+        }
+
+        return EnvVal;
+    }
+
+    auto *RootValue = Env.createValue(Var.getType());
+
+    if (!RootValue) 
+        return nullptr;
+
+    initializeRootValue(*RootValue, Env);
+
+    setLValue(Var, *RootValue, Env);
+
+    return RootValue;
+}
+
+void matchDereferenceExpr(const Stmt *stmt,
+                          const MatchFinder::MatchResult &Result,
+                          Environment &Env) {
+    const auto *Var = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(Var != nullptr);
+
+    // FIXME: Optimize
+    auto *RootValue = getValue(*Var, Env);
+    if (!RootValue) {
+       return;
+    }
+
+    Env.addToFlowCondition(Env.arena().makeNot(getVal(kIsNull, *RootValue).formula()));
+}
+
+void matchCastExpr(const CastExpr *cond, const MatchFinder::MatchResult &Result,
+                   ReverseNullModel::TransferArgs &Data) {
+    auto IsNonnullBranch = Data.Branch;
+    auto &Env = Data.Env;
+
+    const auto *Var = Result.Nodes.getNodeAs<Expr>(kVar);
+    assert(Var != nullptr);
+
+    auto *RootValue = getValue(*Var, Env);
+    if (!RootValue) {
+       return;
+    }
+
+    auto *NewRootValue = Env.createValue(Var->getType());
+    if (!NewRootValue) 
+        return;
+
+    setLValue(*Var, *NewRootValue, Env);
+
+    Arena &A = Env.arena();
+
+    if (IsNonnullBranch) {
+        BoolValue &IsNonnull = getVal(kIsNonnull, *RootValue);
+        BoolValue &IsNull = getVal(kIsNull, *RootValue);
```

Comment:
These local variables and the same local variables in the `else` branch are initialised the same way. Is this deliberate, or a mistake and the initialisation should also be flipped in the `else`?

If not, should we just initialise these references before the `if`?